### PR TITLE
Terminate non-detached test devices on `flutter run` completion

### PR DESCRIPTION
**Pull request by matanlurey**
See https://github.com/flutter/flutter/issues/159154.
See https://github.com/flutter/flutter/pull/159169.

Before this PR, it appeared we were accidentally leaking (keeping active) `flutter_tester` instances (or any test device) after `flutter run` completion, even if the runner was not explicitly detached. I _think_ this is a bug, but I'll check with the tools team and possibly @jonahwilliams before finalizing this.

/cc @jason-simmons

### CODE REVIEW

File: packages/flutter_tools/test/general.shard/run_hot_test.dart [hunk 1]
```diff
@@ -11,11 +16,13 @@ import 'package:test/fake.dart';
...
 
-//import '../src/context.dart';
 import '../src/common.dart';
+import '../src/context.dart';
+import 'hot_shared.dart';
 
 void main() {
-  testWithoutContext('defaultReloadSourcesHelper() handles empty DeviceReloadReports)', () {
+  testWithoutContext(
+      'defaultReloadSourcesHelper() handles empty DeviceReloadReports)', () {
```
Comment (andrewkolos) [inline packages/flutter_tools/test/general.shard/run_hot_test.dart:25]: Nit: formatting

File: packages/flutter_tools/test/general.shard/run_hot_test.dart [hunk 2]
```diff
@@ -29,6 +36,70 @@ void main() {
       const NoOpAnalytics(),
     );
   });
+
+  group('signal handling', () {
+    late final _FakeHotCompatibleFlutterDevice flutterDevice;
+
+    setUpAll(() {
+      flutterDevice = _FakeHotCompatibleFlutterDevice(FakeDevice());
+    });
```
Comment (andrewkolos) [inline packages/flutter_tools/test/general.shard/run_hot_test.dart:45]: Nit: while I don't foresee this `group` growing to encompass a large number of tests, I prefer each test to do its own setup except in scenarios where setup is expensive or `testUsingContext` needs to be used.

https://github.com/flutter/flutter/blob/master/docs/contributing/Style-guide-for-Flutter-repo.md#make-each-test-entirely-self-contained

File: packages/flutter_tools/test/general.shard/run_hot_test.dart [hunk 3]
```diff
@@ -29,6 +36,70 @@ void main() {
...
+            flutterDevice,
+          ],
+          target: 'main.dart',
+          debuggingOptions: DebuggingOptions.disabled(BuildInfo.debug),
+          analytics: _FakeAnalytics(),
+        );
+
+        await runner.run();
+        await runner.cleanupAfterSignal();
+        expect(flutterDevice.wasExited, true);
```
Comment (andrewkolos) [inline packages/flutter_tools/test/general.shard/run_hot_test.dart:67]:
```suggestion
        expect(flutterDevice.wasExited, isTrue);
```

File: packages/flutter_tools/test/general.shard/run_hot_test.dart [hunk 4]
```diff
@@ -29,6 +36,70 @@ void main() {
...
+          ],
+          target: 'main.dart',
+          debuggingOptions: DebuggingOptions.disabled(BuildInfo.debug),
+          analytics: _FakeAnalytics(),
+        );
+
+        await runner.run();
+        await runner.detach();
+        await runner.cleanupAfterSignal();
+        expect(flutterDevice.wasExited, false);
```
Comment (andrewkolos) [inline packages/flutter_tools/test/general.shard/run_hot_test.dart:90]:
```suggestion
        expect(flutterDevice.wasExited, isFalse);
```

File: packages/flutter_tools/test/general.shard/run_hot_test.dart (hunk 2 above)
Comment (andrewkolos) [reply packages/flutter_tools/test/general.shard/run_hot_test.dart:45]: edit: I'm dumb, we are using `testUsingContext`

**Review by andrewkolos — DISMISSED** [review]
LGTM

**Review by jason-simmons — DISMISSED** [review]
I suspect that the original goal of `HotRunner.cleanupAfterSignal` was that a Ctrl-C of flutter_tools should not try to stop the app if the user connected to an already running app using `flutter attach`.

But it looks like `cleanupAfterSignal` intended to have Ctrl-C stop the app if the app was launched by `flutter run`.  However, that does not actually happen because `HotRunner.run` calls `HotRunner.attach`, which sets the `_didAttach` flag.  So a Ctrl-C goes down the `appFinished` code path instead of `exitApp` for both `flutter attach` and `flutter run`.

Having Ctrl-C stop the app on a `flutter run` would be similar to the behavior of the flutter_tools quit command.  Pressing `q` stops the app after a `flutter run` but not a `flutter attach`.

Can anyone confirm whether Ctrl-C is expected to stop an app launched by `flutter run`?

If not, then it may be simpler to always have `cleanupAfterSignal` call `appFinished` and remove the `exitApp` code path.

**Comment by jason-simmons** [comment]
Alternatively, it might make sense to have a flag that represents whether the app connection was initiated through `flutter run` or `flutter attach`.

`cleanupAfterSignal` could then use that flag to decide whether to call `appFinished` or `exitApp`.  (I'm guessing that this is what `_didAttach` was originally supposed to represent).

With that change, Ctrl-C will stop an app that was launched with `flutter run`.  And that will also accomplish the goal of having flutter_tools stop the flutter_tester process if the tool was launched with `flutter run -d flutter-tester` and is then killed by a signal.

**Comment by andrewkolos** [comment]
> Can anyone confirm whether Ctrl-C is expected to stop an app launched by `flutter run`?

I don't believe this is documented. In fact, the current behavior differs based on target platform. When running a Flutter app on an Android emulator, Ctrl-C leaves the app running (unlike sending `q` which exits it). When running a Flutter app on desktop (macOS), Ctrl-C exits the app.

I don't attach/detach when working on Flutter apps personally, so I don't know if my opinion is worth anything. _Personally_, if anything I would expect Ctrl-C to behave similarly to sending `q`. However, `q` also behaves ambiguously. `q` does not quit the app during `flutter attach`, as Jason mentioned, even though the description text still says the app will be. Regardless, this can be treated as a separate issue (as it already is, see https://github.com/flutter/flutter/issues/122303).

**Comment by andrewkolos** [comment]
TL;DR [Jason's last comment](https://github.com/flutter/flutter/pull/159170#issuecomment-2487030104) sounds good to me

File: packages/flutter_tools/test/general.shard/run_hot_test.dart (hunk 3 above)
Comment (matanlurey) [reply packages/flutter_tools/test/general.shard/run_hot_test.dart:67]: (Here and below) My 2c: I don't think there is any value in this suggestion and we should stop making the suggestion - even the pkg/test authors are debating deprecating `isTrue` and `isFalse` because it has no particular value.

If you feel strongly about it, I can make the change anyway.

**Comment by matanlurey** [comment]
> Alternatively, it might make sense to have a flag that represents whether the app connection was initiated through `flutter run` or `flutter attach`.
> 
> `cleanupAfterSignal` could then use that flag to decide whether to call `appFinished` or `exitApp`. (I'm guessing that this is what `_didAttach` was originally supposed to represent).
> 
> With that change, Ctrl-C will stop an app that was launched with `flutter run`. And that will also accomplish the goal of having flutter_tools stop the flutter_tester process if the tool was launched with `flutter run -d flutter-tester` and is then killed by a signal.

Sounds good, moving this back to draft - will ping when updated.

File: packages/flutter_tools/test/general.shard/run_hot_test.dart (hunk 3 above)
Comment (andrewkolos) [reply packages/flutter_tools/test/general.shard/run_hot_test.dart:67]: SGTM. I just checked the output of these matchers when they fail and it's basically the same as not using the matcher.

> even the pkg/test authors are debating deprecating `isTrue` and `isFalse` because it has no particular value

This is news to me.

**Comment by matanlurey** [comment]
Ok, this simplifies things a bit and added another test case!

**Review by jason-simmons — COMMENTED** [review]
Overall LGTM

File: packages/flutter_tools/lib/src/resident_runner.dart [hunk 5]
```diff
@@ -1118,6 +1118,15 @@ abstract class ResidentRunner extends ResidentHandlers {
...
   }
 
+  /// Whether the app being instrumented by the runner is detached.
+  ///
+  /// A detached app can happen one of two ways:
+  /// - [run] is used, and then the created application is manually [detach]ed;
+  /// - [attach] is used to explicitly connect to an already running app.
+  @protected
+  @visibleForTesting
+  bool isDetached = false;
```
[the remark this answers is not in the document]
Comment (matanlurey) [reply packages/flutter_tools/lib/src/resident_runner.dart:1128]: SGTM, will do before merging. I like `stopAppDuringCleanup`.

[the remark this answers is not in the document]
Comment (andrewkolos) [reply packages/flutter_tools/lib/src/resident_runner.dart]: `_attachWithoutMarkingDetached` could probably use a rename then, as well. I'm trying to think of a better name...

[the remark this answers is not in the document]
Comment (andrewkolos) [reply packages/flutter_tools/lib/src/resident_runner.dart]: `_attachWithoutStoppingAppOnDetach`?

[the remark this answers is not in the document]
Comment (matanlurey) [reply packages/flutter_tools/lib/src/resident_runner.dart]: Done.

[the remark this answers is not in the document]
Comment (andrewkolos) [reply packages/flutter_tools/lib/src/resident_runner.dart]: I tweaked again in [rename `_attachWithoutStoppingAppOnDetach` to `_attachWithoutStoppingAppDuringCleanup`](https://github.com/flutter/flutter/pull/159170/commits/1b6dedbbc1ad69b63c83680d021c0eff9885545f). A little bit more verbose but precise—I think.

**Review by andrewkolos — APPROVED** [review]
LGTM

File: packages/flutter_tools/lib/src/resident_runner.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (matanlurey) [reply packages/flutter_tools/lib/src/resident_runner.dart]: Works for me.

[the remark this answers is not in the document]
Comment (jason-simmons) [reply packages/flutter_tools/lib/src/resident_runner.dart]: Or perhaps name it `_attach` or `_attachToApp` or `_connectToApp`?

This method does the work of attaching to the app that is common between `flutter run` and `flutter attach`.

I think `_attachWithoutStoppingAppDuringCleanup` makes it sound like it clears the `stopAppDuringCleanup` flag, which is not the case.

[the remark this answers is not in the document]
Comment (andrewkolos) [reply packages/flutter_tools/lib/src/resident_runner.dart]: Tweaked again in [rename `_attachWithoutStoppingAppDuringCleanup` to `_attach`](https://github.com/flutter/flutter/pull/159170/commits/245a2faf4141927bd15f1decd1046fbd9f127ecc)

**Review by jason-simmons — APPROVED** [review]

File: packages/flutter_tools/lib/src/run_hot.dart [hunk 6]
```diff
@@ -1239,7 +1250,7 @@ class HotRunner extends ResidentRunner {
   Future<void> cleanupAfterSignal() async {
     await stopEchoingDeviceLog();
     await hotRunnerConfig!.runPreShutdownOperations();
-    if (_didAttach) {
+    if (stopAppDuringCleanup) {
```
Comment (jason-simmons) [inline packages/flutter_tools/lib/src/run_hot.dart:1253]: Just noticed that `stopAppDuringCleanup` is inverted.

It should be set to `true` only in the `flutter run` case, and if it is `true` then the cleanup should call `exitApp`.

**Review by jason-simmons — CHANGES_REQUESTED** [review]

File: packages/flutter_tools/test/general.shard/run_hot_test.dart [hunk 7]
```diff
@@ -29,6 +36,92 @@ void main() {
       const NoOpAnalytics(),
     );
   });
+
+  group('signal handling', () {
+    late final _FakeHotCompatibleFlutterDevice flutterDevice;
+
+    setUpAll(() {
+      flutterDevice = _FakeHotCompatibleFlutterDevice(FakeDevice());
```
Comment (jason-simmons) [inline packages/flutter_tools/test/general.shard/run_hot_test.dart:44]: This test group is failing when I run it locally in the default non-randomized order.

`flutterDevice` should be reconstructed for each test case in `setUp`

**Review by jason-simmons — APPROVED** [review]

**Comment by jason-simmons** [comment]
Updated the commit message to keep https://github.com/flutter/flutter/issues/159154 open.

Something like https://github.com/flutter/flutter/pull/159169 will still be needed on Windows.  The termination signal handler will not run on Windows in response to the integration test calling `Process.kill()` on the flutter_tools process.
